src/convert/fasta.rs

```diff
@@ -0,0 +1,379 @@
+//! Convert twobit files to Fasta and vice-versa
+//!
+//! This modules assumes Fasta files to be simple: No comments, no UIPAC stuff.
+//! Just `ACGTacgtN` and `>header lines`. Soft blocks (lower case nucleotides)
+//! will be automatically detected.
+
+use std::borrow::Cow;
+use std::collections::HashMap;
+use std::convert::TryInto;
+use std::fs::File;
+use std::io::{BufReader, Cursor, Read, Seek, SeekFrom, Write};
+use std::path::{Path, PathBuf};
+
+use crate::block::Block;
+use crate::convert::{Nucleotides, SequenceLength, SequenceRead};
+use crate::error::{Error, Result};
+use crate::nucleotide::Nucleotide;
+use crate::TwoBitFile;
+
+type FilePos = u64;
+trait FileAccess: Read + Seek {}
+
+impl FileAccess for BufReader<File> {}
+impl FileAccess for Cursor<Vec<u8>> {}
+
+/// Read a Fasta file and make the extracted information available
+pub struct FastaReader {
+    file_path: Option<PathBuf>,
+    in_memory_data: Option<Cursor<Vec<u8>>>,
+    soft_blocks: HashMap<String, Vec<Block>>,
+    hard_blocks: HashMap<String, Vec<Block>>,
+    sequence_starts: HashMap<String, FilePos>,
+    sequence_lengths: Vec<SequenceLength>,
+}
+
+impl FastaReader {
+    /// Read data from a Fasta file that is stored in a `Vec<u8>`.
+    pub fn mem_open(data: Vec<u8>) -> Result<Self> {
+        let mut reader = Cursor::new(data);
+        let mut result = Self::parse_fasta_file(&mut reader)?;
+        result.in_memory_data = Some(reader);
+        Ok(result)
+    }
+
+    /// Open a Fasta file on disk
+    pub fn open<P: AsRef<Path>>(file_path: P) -> Result<Self> {
+        let fd = File::open(&file_path)?;
+        let mut reader = BufReader::new(fd);
+        let mut result = Self::parse_fasta_file(&mut reader)?;
+        result.file_path = Some(PathBuf::from(file_path.as_ref()));
+        Ok(result)
+    }
+
+    /// Extract all the necessary information from the Fasta file
+    #[allow(clippy::too_many_lines)]
```

Comment:
Do you think this function should be restructured? There are a lot of local variables that keep track of various things and that are needed all over the place.